Add dot product test.

prisae opened the pull request:
## Adds the dot product test to the test suite

I was thinking of adding the dot test to `emg3d`, but thought that maybe it lives better in `discretize`, alongside the `OrderTest` and `check_derivative`. As such it is available to any tool that uses `discretize`, including `SimPEG`.

In `SimPEG` there are dozens of places where the dot-test is defined. I think it would be better to only define it once, here in `discretize`, and use always the same definition (avoiding code repetitions and slightly different implementations). However, adding it in `discretize` means that we would probably have to wait at least for 2 minor releases before we can use it in `SimPEG`.

I draw a lot of inspiration from the implementation in PyLops, and we are reworking the `dottest` over there currently (https://github.com/PyLops/pylops/pull/335, https://github.com/PyLops/pylops/pull/336).

Suggestions for further improvements or adaption for `SimPEG`/`discretize`-language/habits/conventions are welcomed!


### Open questions to discuss

- [ ] Defaults for `rtol`, `atol`?
- [ ] Should we have `shape` and `order` variables? Alternatively, we could only support size, and if a particular shape/order is required one has to wrap it like
   ```
   x, y = 10, 20
   dottest(
       fwd=lamda u: my_fwd_op(u.reshape((x, y), order='F')).ravel('F'),
       ...
       fwd_size=x*y,
       ...
   )
   ```
   So it would be possible even when just having `fwd_size` instead of `fwd_shape`, and no `fwd_order` (all the same applies of course to `ajd_...`). But having it inside the functions makes the `dottest` a bit easier to use for various scenarios.
- [ ] Default order is set to `"C"` - would it, with `discretize`, make more sense to set the default order to `"F"`? Or are most operators on 1D vectors anyway?

jcapriot commented on discretize/tests.py:
```diff
@@ -633,6 +637,110 @@ def Quadratic(x, return_g=True, return_H=True):
+        rhs = np.dot(adj_v.real, u.real) + np.dot(adj_v.imag, u.imag)
+
+    # Check if they are the same.
+    passed = np.isclose(rhs, lhs, rtol, atol)
```
probably just use `np.testing.assert_allclose(rhs, lhs, rtol, atol)`

This function works well within testing environments, and will automatically raise correct errors, with much more intuitive error messages.

You can also pass it an `err_msg` kwarg to print when it fails as well. So maybe something like "Dot test failed".

prisae replied:
Implemented.

jcapriot commented on discretize/tests.py:
```diff
@@ -633,6 +637,110 @@ def Quadratic(x, return_g=True, return_H=True):
+    passed = np.isclose(rhs, lhs, rtol, atol)
+
+    # Verbosity and error.
+    if (not passed and raise_error) or verb:
```
And then here, just remove the rest of the function.  Which, I believe would remove the use of the `verb` parameter.

prisae replied:
Goes to the main discussion - should it be a check or an assertion.

prisae replied:
I removed the `verb` parameter, and renamed `raise_error` to `assert_error` (True by default). As such, I kept the "verbose" part so it could either be used as an assertion in tests, or as whatever the users want it to returning the boolean test result.

jcapriot commented on tests/base/test_tests.py:
```diff
@@ -0,0 +1,67 @@
+        P = discretize.utils.volume_average(mesh1, mesh2)
+
+        # return
+        assert dottest(
```
if dottest raises the proper assertions (when switching to `np.testing.assert_allclose`, then we do not need to `assert` the dot test.

prisae replied:
Done.

jcapriot commented on discretize/tests.py:
```diff
@@ -633,6 +637,110 @@ def Quadratic(x, return_g=True, return_H=True):
 
 
+# TESTS
+def dottest(
```
how about we rename this function to `assert_isadjoint` ? Maybe follow the `np.testing` naming scheme for it.

prisae replied:
Done, good idea.

jcapriot reviewed (commented):
Left some review comments here

prisae commented:
Thanks @jcapriot for the feedback, I will have to think about them. I thought about it more like a user-function, that can be used in general. Your comments go more towards it being a testing utility, using the testing convention/things. I like it, makes probably sense!

prisae commented:
Just to check @jcapriot - the `OrderTest` and the `check_derivative` in the same file are verbose, and do not `assert`, but check. I thought of the `dottest` to be a similar routine, similarly used. Are the plans to change the `OrderTest` & `check_derivative` to also more resemble `np.testing/assertion` routines?

prisae commented on discretize/tests.py:
```diff
@@ -633,6 +637,110 @@ def Quadratic(x, return_g=True, return_H=True):
+    v = random(np.product(adj_shape), adj_complex)
+
+    # Carry out dot product test.
+    fwd_u = forward(u.reshape(fwd_shape, order=fwd_order)).ravel(fwd_order)
```
This won't work if the `forward` and the `adjoint` take different shapes.

jcapriot replied:
As long as the out put of the forward matches adj_shape, and the output of adj match’s forward shape. It will work. The two functions do not need to accept the same shaped argument.

prisae replied:
This is/should be true indeed. I had a test example where it failed, but probably my example was bonkers! I'll look at it again.

jcapriot replied:
If the functions returned the appropriate shaped arrays, you could even flatten them out irrespective of their internal ordering, as long as they're both flattened out with the same ordering.

jcapriot commented:
I do not like the way `check_derivative` and `OrderTest` are currently used for tests. When they fail, the only output you get is "AssertionError: False is not True" which is horribly nondescript. To find out why they actually failed, you need to search through the entire output of `pytest -s -v`.

If all you want for development testing environment is for it to write out the result of wTJv and (vJ)Tw, then simply add a print statement comparing the two before `np.testing.assert_allclose`

jcapriot commented:
Also the names `fwd_shape` and `adj_shape` are slightly too nondescript in my opinion. Do they mean "what is the shape of the matrix represented by this function?" because how they are used answers the question: "What is the output shape of this function?" Maybe change their names to `fwd_in_shape` and `fwd_out_shape`? Then by definition the adjoint would accept a `fwd_out_shape` and return a `fwd_in_shape`.

prisae commented:
Thanks for the clarification regarding `OrderTest` and `check_derivative`, that makes things clearer. I implemented most of your suggestions. I further simplified `fwd_in_shape and `fwd_out_shape` to simply `shape_in` and `shape_out`; accordingly, I renamed `fwd_complex` and `adj_complex` to `complex_in` and `complex_out`.

Regarding assertion and verbosity: Have a look, I sort of made a compromise. By default, it acts now as a regular assertion test: `assert_isajdoint()` will be silent if passed, or raise the proper error if failed, using `np.testing.assert_allclose()`. However, if `assert_error` is set to False, it will instead print a message and return the test result as boolean. So it can still be used in other ways than pure testing. Would that be OK?

prisae commented:
The alternative would be to call it `shape_u`/`shape_v` and `complex_u`/`complex_v` to follow the docstring and code, instead of `_in`/`_out`.

jcapriot commented on discretize/tests.py:
```diff
@@ -633,6 +637,121 @@ def Quadratic(x, return_g=True, return_H=True):
+
+    Parameters
+    ----------
+    forward : function
```
Forward and adjoint should be `callable` not function.

prisae replied:
Yes! I knew `function` was wrong, but I couldn't remember what it should be. Thanks.

jcapriot commented on discretize/tests.py:
```diff
@@ -633,6 +637,121 @@ def Quadratic(x, return_g=True, return_H=True):
+
+    Returns
+    -------
+    passed : bool, returned if assert_error=False
```
this line should only have the dtype with it, put any descriptions on the next line

```
passed : bool, optional
    Result of the dot product test. returned if `assert_error` is False.
```

prisae replied:
OK

jcapriot reviewed (commented):
Looks good to me @prisae, just a few minor changes on the docstring, then good to go.
